File: problems/array/subarray_sum_equals_k.py
from typing import List
from collections import defaultdict


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res = 0
        pref = 0
        mem = defaultdict(int)
        mem[0] = 1
        for x in nums:
            pref += x
            if pref - k in mem:
                res += mem[pref - k]

            mem[pref] += 1

        return res


if __name__ == "__main__":
    assert Solution().subarraySum([1, 1, 1], k=2) == 2
    assert Solution().subarraySum([1, 2, 3], k=3) == 2

File: problems/array/subarray_sum_equals_k.go
package main

func subarraySum(nums []int, k int) int {
	var prefix int
	var res int
	mem := map[int]int{}
	mem[0] = 1

	for i := 0; i < len(nums); i++ {
		prefix = prefix + nums[i]
		d := prefix - k
		count, ok := mem[d]
		if ok {
			res = res + count
		}

		cur, ok := mem[prefix]
		if !ok {
			mem[prefix] = 1
		} else {
			mem[prefix] = cur + 1
		}
	}
	return res
}
